### ISSUE TITLE
list_objects_v2 fails with SignatureDoesNotMatch when prefix contains '{' or '}'

### ISSUE THREAD

**Issue by eagletmt** [awslabs/aws-sdk-rust#330]
### Version

```
│   ├── aws-sdk-sts v0.2.0
├── aws-sdk-s3 v0.2.0
```

### Platform

```
Linux reinforce 5.4.72-microsoft-standard-WSL2 #1 SMP Wed Oct 28 23:40:43 UTC 2020 x86_64 GNU/Linux
```
### AWS Services

S3

### Description

list_objects_v2 fails with SignatureDoesNotMatch when prefix contains '{' or '}'.
The situation is similar to #120 .

I tried this code:

```rust
#[tokio::main]
async fn main() {
    tracing_subscriber::fmt::init();

    let shared_config = aws_config::load_from_env().await;
    let client = aws_sdk_s3::Client::new(&shared_config);

    for prefix in ["{foo", "bar}"] {
        let resp = client
            .list_objects_v2()
            .bucket("wanko-cc")
            .prefix(prefix)
            .send()
            .await;
        if let Err(e) = resp {
            eprintln!("{}", e);
        }
    }
}
```

I expected to see this happen: No errors reported

Instead, this happened:

```
Dec 09 12:49:41.244 TRACE aws_smithy_http_tower::dispatch: request=Request { method: GET, uri: https://s3.ap-northeast-1.amazonaws.com/wanko-cc?list-type=2&prefix={foo, version: HTTP/1.1, headers: {"user-agent": "aws-sdk-rust/0.2.0 os/linux lang/rust/1.57.0", "x-amz-user-agent": "aws-sdk-rust/0.2.0 api/s3/0.2.0 os/linux lang/rust/1.57.0", "x-amz-date": "20211209T034941Z", "authorization": Sensitive, "x-amz-content-sha256": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"}, body: SdkBody { inner: Once(Some(b"")), retryable: true } }
Dec 09 12:49:41.293 TRACE aws_smithy_http::middleware: http_response=Response { status: 403, version: HTTP/1.1, headers: {"x-amz-bucket-region": "ap-northeast-1", "x-amz-request-id": "XBTVMATD57E7Q2JQ", "x-amz-id-2": "QGEB5QU2yN6BzzLMmi7Lu1uOoILn1Ez/2EvvLwx7HtWbBg1ELTIp/vdLxwSQllz6mFKB1+WrqdU=", "content-type": "application/xml", "transfer-encoding": "chunked", "date": "Thu, 09 Dec 2021 03:49:42 GMT", "server": "AmazonS3"}, body: b"<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>SignatureDoesNotMatch</Code><Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message><AWSAccessKeyId>AKIA(masked)</AWSAccessKeyId><StringToSign>AWS4-HMAC-SHA256\n20211209T034941Z\n20211209/ap-northeast-1/s3/aws4_request\n3094659af5dc6901dca2a3381ea11f91fb6d7a7cef4203d1691e2a58d4a5cea1</StringToSign><SignatureProvided>f31ed2ae6c5c1e0bc2e5917aa202c278491bc7cf5cfdc650e2115fe1bfd52c8c</SignatureProvided><StringToSignBytes>41 57 53 34 2d 48 4d 41 43 2d 53 48 41 32 35 36 0a 32 30 32 31 31 32 30 39 54 30 33 34 39 34 31 5a 0a 32 30 32 31 31 32 30 39 2f 61 70 2d 6e 6f 72 74 68 65 61 73 74 2d 31 2f 73 33 2f 61 77 73 34 5f 72 65 71 75 65 73 74 0a 33 30 39 34 36 35 39 61 66 35 64 63 36 39 30 31 64 63 61 32 61 33 33 38 31 65 61 31 31 66 39 31 66 62 36 64 37 61 37 63 65 66 34 32 30 33 64 31 36 39 31 65 32 61 35 38 64 34 61 35 63 65 61 31</StringToSignBytes><CanonicalRequest>GET\n/wanko-cc\nlist-type=2&amp;prefix=%7Bfoo\nhost:s3.ap-northeast-1.amazonaws.com\nx-amz-content-sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date:20211209T034941Z\nx-amz-user-agent:aws-sdk-rust/0.2.0 api/s3/0.2.0 os/linux lang/rust/1.57.0\n\nhost;x-amz-content-sha256;x-amz-date;x-amz-user-agent\ne3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855</CanonicalRequest><CanonicalRequestBytes>47 45 54 0a 2f 77 61 6e 6b 6f 2d 63 63 0a 6c 69 73 74 2d 74 79 70 65 3d 32 26 70 72 65 66 69 78 3d 25 37 42 66 6f 6f 0a 68 6f 73 74 3a 73 33 2e 61 70 2d 6e 6f 72 74 68 65 61 73 74 2d 31 2e 61 6d 61 7a 6f 6e 61 77 73 2e 63 6f 6d 0a 78 2d 61 6d 7a 2d 63 6f 6e 74 65 6e 74 2d 73 68 61 32 35 36 3a 65 33 62 30 63 34 34 32 39 38 66 63 31 63 31 34 39 61 66 62 66 34 63 38 39 39 36 66 62 39 32 34 32 37 61 65 34 31 65 34 36 34 39 62 39 33 34 63 61 34 39 35 39 39 31 62 37 38 35 32 62 38 35 35 0a 78 2d 61 6d 7a 2d 64 61 74 65 3a 32 30 32 31 31 32 30 39 54 30 33 34 39 34 31 5a 0a 78 2d 61 6d 7a 2d 75 73 65 72 2d 61 67 65 6e 74 3a 61 77 73 2d 73 64 6b 2d 72 75 73 74 2f 30 2e 32 2e 30 20 61 70 69 2f 73 33 2f 30 2e 32 2e 30 20 6f 73 2f 6c 69 6e 75 78 20 6c 61 6e 67 2f 72 75 73 74 2f 31 2e 35 37 2e 30 0a 0a 68 6f 73 74 3b 78 2d 61 6d 7a 2d 63 6f 6e 74 65 6e 74 2d 73 68 61 32 35 36 3b 78 2d 61 6d 7a 2d 64 61 74 65 3b 78 2d 61 6d 7a 2d 75 73 65 72 2d 61 67 65 6e 74 0a 65 33 62 30 63 34 34 32 39 38 66 63 31 63 31 34 39 61 66 62 66 34 63 38 39 39 36 66 62 39 32 34 32 37 61 65 34 31 65 34 36 34 39 62 39 33 34 63 61 34 39 35 39 39 31 62 37 38 35 32 62 38 35 35</CanonicalRequestBytes><RequestId>XBTVMATD57E7Q2JQ</RequestId><HostId>QGEB5QU2yN6BzzLMmi7Lu1uOoILn1Ez/2EvvLwx7HtWbBg1ELTIp/vdLxwSQllz6mFKB1+WrqdU=</HostId></Error>" }
Error { code: "SignatureDoesNotMatch", message: "The request signature we calculated does not match the signature you provided. Check your key and signing method.", request_id: "XBTVMATD57E7Q2JQ", s3_extended_request_id: "QGEB5QU2yN6BzzLMmi7Lu1uOoILn1Ez/2EvvLwx7HtWbBg1ELTIp/vdLxwSQllz6mFKB1+WrqdU=" }
Dec 09 12:49:41.294 TRACE aws_smithy_http_tower::dispatch: request=Request { method: GET, uri: https://s3.ap-northeast-1.amazonaws.com/wanko-cc?list-type=2&prefix=bar}, version: HTTP/1.1, headers: {"user-agent": "aws-sdk-rust/0.2.0 os/linux lang/rust/1.57.0", "x-amz-user-agent": "aws-sdk-rust/0.2.0 api/s3/0.2.0 os/linux lang/rust/1.57.0", "x-amz-date": "20211209T034941Z", "authorization": Sensitive, "x-amz-content-sha256": "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"}, body: SdkBody { inner: Once(Some(b"")), retryable: true } }
Dec 09 12:49:41.306 TRACE aws_smithy_http::middleware: http_response=Response { status: 403, version: HTTP/1.1, headers: {"x-amz-bucket-region": "ap-northeast-1", "x-amz-request-id": "XBTMEKWP50WSHAR2", "x-amz-id-2": "taRr5S6i6rBbkUrHR/oX/sKWkLM3ZjMNyLToJ7UwYLBDy0uPdbh/TaoOqMehszE7lLWth9tF+Ds=", "content-type": "application/xml", "transfer-encoding": "chunked", "date": "Thu, 09 Dec 2021 03:49:42 GMT", "server": "AmazonS3"}, body: b"<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<Error><Code>SignatureDoesNotMatch</Code><Message>The request signature we calculated does not match the signature you provided. Check your key and signing method.</Message><AWSAccessKeyId>AKIA(masked)</AWSAccessKeyId><StringToSign>AWS4-HMAC-SHA256\n20211209T034941Z\n20211209/ap-northeast-1/s3/aws4_request\n9cb24d17c285185ba2577bfb0a52ad9f20b64d39c6786f9fe3be81adfd8aa755</StringToSign><SignatureProvided>5dfee0d666208c9a2985595aa6d021825fd07259474b850098f87054636b8a0d</SignatureProvided><StringToSignBytes>41 57 53 34 2d 48 4d 41 43 2d 53 48 41 32 35 36 0a 32 30 32 31 31 32 30 39 54 30 33 34 39 34 31 5a 0a 32 30 32 31 31 32 30 39 2f 61 70 2d 6e 6f 72 74 68 65 61 73 74 2d 31 2f 73 33 2f 61 77 73 34 5f 72 65 71 75 65 73 74 0a 39 63 62 32 34 64 31 37 63 32 38 35 31 38 35 62 61 32 35 37 37 62 66 62 30 61 35 32 61 64 39 66 32 30 62 36 34 64 33 39 63 36 37 38 36 66 39 66 65 33 62 65 38 31 61 64 66 64 38 61 61 37 35 35</StringToSignBytes><CanonicalRequest>GET\n/wanko-cc\nlist-type=2&amp;prefix=bar%7D\nhost:s3.ap-northeast-1.amazonaws.com\nx-amz-content-sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date:20211209T034941Z\nx-amz-user-agent:aws-sdk-rust/0.2.0 api/s3/0.2.0 os/linux lang/rust/1.57.0\n\nhost;x-amz-content-sha256;x-amz-date;x-amz-user-agent\ne3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855</CanonicalRequest><CanonicalRequestBytes>47 45 54 0a 2f 77 61 6e 6b 6f 2d 63 63 0a 6c 69 73 74 2d 74 79 70 65 3d 32 26 70 72 65 66 69 78 3d 62 61 72 25 37 44 0a 68 6f 73 74 3a 73 33 2e 61 70 2d 6e 6f 72 74 68 65 61 73 74 2d 31 2e 61 6d 61 7a 6f 6e 61 77 73 2e 63 6f 6d 0a 78 2d 61 6d 7a 2d 63 6f 6e 74 65 6e 74 2d 73 68 61 32 35 36 3a 65 33 62 30 63 34 34 32 39 38 66 63 31 63 31 34 39 61 66 62 66 34 63 38 39 39 36 66 62 39 32 34 32 37 61 65 34 31 65 34 36 34 39 62 39 33 34 63 61 34 39 35 39 39 31 62 37 38 35 32 62 38 35 35 0a 78 2d 61 6d 7a 2d 64 61 74 65 3a 32 30 32 31 31 32 30 39 54 30 33 34 39 34 31 5a 0a 78 2d 61 6d 7a 2d 75 73 65 72 2d 61 67 65 6e 74 3a 61 77 73 2d 73 64 6b 2d 72 75 73 74 2f 30 2e 32 2e 30 20 61 70 69 2f 73 33 2f 30 2e 32 2e 30 20 6f 73 2f 6c 69 6e 75 78 20 6c 61 6e 67 2f 72 75 73 74 2f 31 2e 35 37 2e 30 0a 0a 68 6f 73 74 3b 78 2d 61 6d 7a 2d 63 6f 6e 74 65 6e 74 2d 73 68 61 32 35 36 3b 78 2d 61 6d 7a 2d 64 61 74 65 3b 78 2d 61 6d 7a 2d 75 73 65 72 2d 61 67 65 6e 74 0a 65 33 62 30 63 34 34 32 39 38 66 63 31 63 31 34 39 61 66 62 66 34 63 38 39 39 36 66 62 39 32 34 32 37 61 65 34 31 65 34 36 34 39 62 39 33 34 63 61 34 39 35 39 39 31 62 37 38 35 32 62 38 35 35</CanonicalRequestBytes><RequestId>XBTMEKWP50WSHAR2</RequestId><HostId>taRr5S6i6rBbkUrHR/oX/sKWkLM3ZjMNyLToJ7UwYLBDy0uPdbh/TaoOqMehszE7lLWth9tF+Ds=</HostId></Error>" }
Error { code: "SignatureDoesNotMatch", message: "The request signature we calculated does not match the signature you provided. Check your key and signing method.", request_id: "XBTMEKWP50WSHAR2", s3_extended_request_id: "taRr5S6i6rBbkUrHR/oX/sKWkLM3ZjMNyLToJ7UwYLBDy0uPdbh/TaoOqMehszE7lLWth9tF+Ds=" }
```

I've masked AWSAccessKeyId, just in case.

**Comment by Velfi** [comment]
Thanks for submitting this, we'll investigate.

**Comment by Velfi** [comment]
This happens because our query paths weren't being encoded in an identical manner to S3. I have a fix and am adding tests to ensure we never have a query encoding issue like this again

**Comment by Velfi** [comment]
Reopening this until we actually ship the fix
